File: src/ansi.js
'use strict';

// Terminal escape-sequence parsing, colour resolution and WCAG contrast.
// Everything here works on the raw bytes a CLI actually wrote, because that is
// the only thing a screen reader or a low-vision user ever gets.

const ESC = '\x1b';
const BEL = '\x07';
const ST = '\x9c';

// xterm's default palette for the 16 named colours. Terminals re-theme these,
// but a CLI that fails contrast against the defaults fails for most people.
const BASE16 = [
  [0, 0, 0], [205, 0, 0], [0, 205, 0], [205, 205, 0],
  [0, 0, 238], [205, 0, 205], [0, 205, 205], [229, 229, 229],
  [127, 127, 127], [255, 0, 0], [0, 255, 0], [255, 255, 0],
  [92, 92, 255], [255, 0, 255], [0, 255, 255], [255, 255, 255],
];

const NAMES = ['black', 'red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'white',
  'bright black', 'bright red', 'bright green', 'bright yellow',
  'bright blue', 'bright magenta', 'bright cyan', 'bright white'];

const CUBE = [0, 95, 135, 175, 215, 255];

function xterm256(n) {
  if (n < 16) return BASE16[n];
  if (n < 232) {
    const i = n - 16;
    return [CUBE[Math.floor(i / 36) % 6], CUBE[Math.floor(i / 6) % 6], CUBE[i % 6]];
  }
  const v = 8 + (n - 232) * 10;
  return [v, v, v];
}

function luminance([r, g, b]) {
  const f = (c) => {
    const s = c / 255;
    return s <= 0.03928 ? s / 12.92 : Math.pow((s + 0.055) / 1.055, 2.4);
  };
  return 0.2126 * f(r) + 0.7152 * f(g) + 0.0722 * f(b);
}

function contrast(a, b) {
  const [x, y] = [luminance(a), luminance(b)].sort((m, n) => n - m);
  return (x + 0.05) / (y + 0.05);
}

function hex(rgb) {
  return '#' + rgb.map((c) => Math.max(0, Math.min(255, Math.round(c))).toString(16).padStart(2, '0')).join('');
}

/** Nearest named colour, for describing an arbitrary RGB in a report. */
function describe(rgb) {
  let best = 0;
  let bestD = Infinity;
  BASE16.forEach((c, i) => {
    const d = (c[0] - rgb[0]) ** 2 + (c[1] - rgb[1]) ** 2 + (c[2] - rgb[2]) ** 2;
    if (d < bestD) { bestD = d; best = i; }
  });
  return bestD === 0 ? NAMES[best] : `${NAMES[best]}-ish ${hex(rgb)}`;
}

// The two backgrounds worth testing against: a typical dark terminal and a
// typical light one. A colour that only works on one of them is a real problem,
// because the CLI cannot know which the user has.
const DARK_BG = [30, 30, 30];
const LIGHT_BG = [255, 255, 255];

function blankStyle() {
  return {
    fg: null, bg: null, bold: false, dim: false, italic: false,
    underline: false, reverse: false, fgName: null, bgName: null,
  };
}

// Apply one SGR (Select Graphic Rendition) sequence to a style.
function applySgr(style, params) {
  const p = params.length ? params : [0];
  for (let i = 0; i < p.length; i++) {
    const n = p[i];
    if (n === 0) Object.assign(style, blankStyle());
    else if (n === 1) style.bold = true;
    else if (n === 2) style.dim = true;
    else if (n === 3) style.italic = true;
    else if (n === 4) style.underline = true;
    else if (n === 7) style.reverse = true;
    else if (n === 22) { style.bold = false; style.dim = false; }
    else if (n === 23) style.italic = false;
    else if (n === 24) style.underline = false;
    else if (n === 27) style.reverse = false;
    else if (n >= 30 && n <= 37) { style.fg = BASE16[n - 30]; style.fgName = String(n); }
    else if (n === 39) { style.fg = null; style.fgName = null; }
    else if (n >= 40 && n <= 47) { style.bg = BASE16[n - 40]; style.bgName = String(n); }
    else if (n === 49) { style.bg = null; style.bgName = null; }
    else if (n >= 90 && n <= 97) { style.fg = BASE16[n - 90 + 8]; style.fgName = String(n); }
    else if (n >= 100 && n <= 107) { style.bg = BASE16[n - 100 + 8]; style.bgName = String(n); }
    else if (n === 38 || n === 48) {
      const target = n === 38 ? 'fg' : 'bg';
      const nameKey = n === 38 ? 'fgName' : 'bgName';
      if (p[i + 1] === 5) {
        style[target] = xterm256(p[i + 2] || 0);
        style[nameKey] = `${n};5;${p[i + 2]}`;
        i += 2;
      } else if (p[i + 1] === 2) {
        style[target] = [p[i + 2] || 0, p[i + 3] || 0, p[i + 4] || 0];
        style[nameKey] = `${n};2;${p[i + 2]};${p[i + 3]};${p[i + 4]}`;
        i += 4;
      }
    }
  }
}

// Many terminals render bold + a basic colour as the bright variant, so the
// colour a CLI gets is often not the one it named.
function effectiveFg(style) {
  if (!style.fg) return null;
  if (style.bold && style.fgName && /^3[0-7]$/.test(style.fgName)) {
    return BASE16[Number(style.fgName) - 30 + 8];
  }
  return style.fg;
}

/**
 * Walk a raw output stream, returning the visible text, the styled runs within
 * it, and a tally of the control sequences that affect assistive tech.
 */
function parse(raw) {
  const style = blankStyle();
  const runs = [];
  let plain = '';
  let current = null;
  const controls = {
    carriageReturns: 0, eraseLine: 0, eraseDisplay: 0, cursorMove: 0,
    cursorHide: 0, cursorShow: 0, altScreenEnter: 0, altScreenLeave: 0,
    bell: 0, osc: 0, sgr: 0,
  };

  const flush = () => {
    if (current && current.text.length) runs.push(current);
    current = null;
  };

  for (let i = 0; i < raw.length; i++) {
    const ch = raw[i];

    if (ch === ESC) {
      const next = raw[i + 1];
      if (next === '[') {
        // CSI: ESC [ params intermediates final
        let j = i + 2;
        let body = '';
        while (j < raw.length && !/[@-~]/.test(raw[j])) { body += raw[j]; j++; }
        const final = raw[j];
        const isPrivate = /^[?<>=]/.test(body);
        const nums = body.replace(/^[?<>=]/, '').split(';')
          .map((s) => (s === '' ? 0 : Number(s)))
          .filter((n) => !Number.isNaN(n));

        if (final === 'm' && !isPrivate) {
          flush();
          applySgr(style, nums);
          controls.sgr++;
        } else if (final === 'K') controls.eraseLine++;
        else if (final === 'J') controls.eraseDisplay++;
        else if (final && 'ABCDEFGHfd'.includes(final)) controls.cursorMove++;
        else if (isPrivate && (final === 'h' || final === 'l')) {
          const mode = nums[0];
          if (mode === 25) { if (final === 'l') controls.cursorHide++; else controls.cursorShow++; }
          if (mode === 1049 || mode === 47 || mode === 1047) {
            if (final === 'h') controls.altScreenEnter++; else controls.altScreenLeave++;
          }
        }
        i = j === undefined ? raw.length : j;
        continue;
      }
      if (next === ']') {
        // OSC: ends at BEL or ST
        let j = i + 2;
        while (j < raw.length && raw[j] !== BEL && raw[j] !== ST
               && !(raw[j] === ESC && raw[j + 1] === '\\')) j++;
        controls.osc++;
        i = raw[j] === ESC ? j + 1 : j;
        continue;
      }
      i++; // some other two-byte escape
      continue;
    }

    if (ch === '\r') { controls.carriageReturns++; plain += ch; continue; }
    if (ch === BEL) { controls.bell++; continue; }

    plain += ch;
    if (ch === '\n') { flush(); continue; }

    if (!current) {
      current = {
        text: '',
        style: { ...style, fg: effectiveFg(style) },
        line: plain.split('\n').length - 1,
      };
    }
    current.text += ch;
  }
  flush();

  return { plain, runs, controls };
}

/** Strip every escape sequence, leaving what a plain-text reader would get. */
function strip(raw) {
  return parse(raw).plain;
}

// Which SGR parameters actually change a colour. Bold, underline and italic are
// emphasis: they survive a monochrome terminal, they carry meaning without it,
// and NO_COLOR is not asking anyone to give them up. Faint and reverse do change
// what colour lands on the screen, so they count.
function isColourParam(n) {
  return (n >= 30 && n <= 49) || (n >= 90 && n <= 97) || (n >= 100 && n <= 107)
    || n === 2 || n === 7;
}

/**
 * True if the stream sets a colour, as opposed to merely emphasising text.
 *
 * The distinction matters more than it looks. A pager renders man-page style
 * help with bold and underline, and those bytes arrive in our capture mixed in
 * with the CLI's own; counting them as colour means accusing a perfectly
 * well-behaved tool of ignoring NO_COLOR.
 */
function hasColour(raw) {
  const matches = raw.match(/\x1b\[[0-9;]*m/g);
  if (!matches) return false;
  return matches.some((seq) => {
    const body = seq.slice(2, -1);
    const params = body === '' ? [0] : body.split(';').map(Number);
    return params.some((n) => !Number.isNaN(n) && isColourParam(n));
  });
}

module.exports = {
  parse, strip, hasColour, isColourParam, contrast, luminance, hex, describe,
  effectiveFg, xterm256, BASE16, NAMES, DARK_BG, LIGHT_BG, ESC,
};

File: internal/ansi/ansi.go
// Package ansi parses terminal escape sequences, resolves colours to RGB and
// computes WCAG contrast. Everything here works on the raw bytes a CLI actually
// wrote, because that is the only thing a screen reader or a low-vision user
// ever gets.
package ansi

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

// RGB is an 8-bit-per-channel colour.
type RGB [3]int

// BASE16 is xterm's default palette for the 16 named colours. Terminals
// re-theme these, but a CLI that fails contrast against the defaults fails for
// most people.
var BASE16 = [16]RGB{
	{0, 0, 0}, {205, 0, 0}, {0, 205, 0}, {205, 205, 0},
	{0, 0, 238}, {205, 0, 205}, {0, 205, 205}, {229, 229, 229},
	{127, 127, 127}, {255, 0, 0}, {0, 255, 0}, {255, 255, 0},
	{92, 92, 255}, {255, 0, 255}, {0, 255, 255}, {255, 255, 255},
}

var names = [16]string{
	"black", "red", "green", "yellow", "blue", "magenta", "cyan", "white",
	"bright black", "bright red", "bright green", "bright yellow",
	"bright blue", "bright magenta", "bright cyan", "bright white",
}

// The two backgrounds worth testing against: a typical dark terminal and a
// typical light one. A colour that only works on one of them is a real problem,
// because the CLI cannot know which the user has.
var (
	DarkBG  = RGB{30, 30, 30}
	LightBG = RGB{255, 255, 255}
)

var cube = [6]int{0, 95, 135, 175, 215, 255}

// XTerm256 resolves a 256-colour index to RGB.
func XTerm256(n int) RGB {
	switch {
	case n < 16:
		return BASE16[n]
	case n < 232:
		i := n - 16
		return RGB{cube[(i/36)%6], cube[(i/6)%6], cube[i%6]}
	default:
		v := 8 + (n-232)*10
		return RGB{v, v, v}
	}
}

// Luminance is the WCAG relative luminance of a colour.
func Luminance(c RGB) float64 {
	f := func(v int) float64 {
		s := float64(v) / 255
		if s <= 0.03928 {
			return s / 12.92
		}
		return math.Pow((s+0.055)/1.055, 2.4)
	}
	return 0.2126*f(c[0]) + 0.7152*f(c[1]) + 0.0722*f(c[2])
}

// Contrast is the WCAG contrast ratio between two colours.
func Contrast(a, b RGB) float64 {
	x, y := Luminance(a), Luminance(b)
	if y > x {
		x, y = y, x
	}
	return (x + 0.05) / (y + 0.05)
}

// Hex renders a colour as #rrggbb.
func Hex(c RGB) string {
	clamp := func(v int) int { return max(0, min(255, v)) }
	return fmt.Sprintf("#%02x%02x%02x", clamp(c[0]), clamp(c[1]), clamp(c[2]))
}

// Describe names the nearest palette colour, for use in a report.
func Describe(c RGB) string {
	best, bestD := 0, math.MaxInt
	for i, p := range BASE16 {
		d := (p[0]-c[0])*(p[0]-c[0]) + (p[1]-c[1])*(p[1]-c[1]) + (p[2]-c[2])*(p[2]-c[2])
		if d < bestD {
			best, bestD = i, d
		}
	}
	if bestD == 0 {
		return names[best]
	}
	return names[best] + "-ish " + Hex(c)
}

// Style is the set of SGR attributes in force for a run of text.
type Style struct {
	FG, BG                            *RGB
	FGName, BGName                    string
	Bold, Dim, Italic, Underline, Rev bool
}

// Run is a stretch of text sharing one style.
type Run struct {
	Text        string
	Style       Style
	Line        int
	Probe       string
	ContextLine string
}

// Controls tallies the sequences that matter to assistive technology.
type Controls struct {
	CarriageReturns, EraseLine, EraseDisplay, CursorMove   int
	CursorHide, CursorShow, AltScreenEnter, AltScreenLeave int
	Bell, OSC, SGR                                         int
}

// Parsed is the result of walking an output stream.
type Parsed struct {
	Plain    string
	Runs     []Run
	Controls Controls
}

func (s *Style) apply(params []int) {
	if len(params) == 0 {
		params = []int{0}
	}
	for i := 0; i < len(params); i++ {
		n := params[i]
		switch {
		case n == 0:
			*s = Style{}
		case n == 1:
			s.Bold = true
		case n == 2:
			s.Dim = true
		case n == 3:
			s.Italic = true
		case n == 4:
			s.Underline = true
		case n == 7:
			s.Rev = true
		case n == 22:
			s.Bold, s.Dim = false, false
		case n == 23:
			s.Italic = false
		case n == 24:
			s.Underline = false
		case n == 27:
			s.Rev = false
		case n >= 30 && n <= 37:
			c := BASE16[n-30]
			s.FG, s.FGName = &c, strconv.Itoa(n)
		case n == 39:
			s.FG, s.FGName = nil, ""
		case n >= 40 && n <= 47:
			c := BASE16[n-40]
			s.BG, s.BGName = &c, strconv.Itoa(n)
		case n == 49:
			s.BG, s.BGName = nil, ""
		case n >= 90 && n <= 97:
			c := BASE16[n-90+8]
			s.FG, s.FGName = &c, strconv.Itoa(n)
		case n >= 100 && n <= 107:
			c := BASE16[n-100+8]
			s.BG, s.BGName = &c, strconv.Itoa(n)
		case n == 38 || n == 48:
			var c RGB
			var name string
			consumed := 0
			if i+1 < len(params) && params[i+1] == 5 && i+2 < len(params) {
				c = XTerm256(params[i+2])
				name = fmt.Sprintf("%d;5;%d", n, params[i+2])
				consumed = 2
			} else if i+1 < len(params) && params[i+1] == 2 && i+4 < len(params) {
				c = RGB{params[i+2], params[i+3], params[i+4]}
				name = fmt.Sprintf("%d;2;%d;%d;%d", n, params[i+2], params[i+3], params[i+4])
				consumed = 4
			} else {
				continue
			}
			if n == 38 {
				s.FG, s.FGName = &c, name
			} else {
				s.BG, s.BGName = &c, name
			}
			i += consumed
		}
	}
}

// EffectiveFG accounts for terminals rendering bold plus a basic colour as the
// bright variant, so the colour a CLI gets is often not the one it named.
func EffectiveFG(s Style) *RGB {
	if s.FG == nil {
		return nil
	}
	if s.Bold && len(s.FGName) == 2 && s.FGName[0] == '3' && s.FGName[1] >= '0' && s.FGName[1] <= '7' {
		n, _ := strconv.Atoi(s.FGName)
		c := BASE16[n-30+8]
		return &c
	}
	return s.FG
}

func isFinal(b byte) bool { return b >= '@' && b <= '~' }

// Parse walks a raw output stream, returning the visible text, the styled runs
// within it, and a tally of the control sequences that affect assistive tech.
func Parse(raw string) Parsed {
	var (
		style   Style
		out     Parsed
		plain   strings.Builder
		cur     *Run
		lineNum int
	)

	flush := func() {
		if cur != nil && cur.Text != "" {
			out.Runs = append(out.Runs, *cur)
		}
		cur = nil
	}

	for i := 0; i < len(raw); i++ {
		ch := raw[i]

		if ch == 0x1b {
			if i+1 < len(raw) && raw[i+1] == '[' {
				j := i + 2
				start := j
				for j < len(raw) && !isFinal(raw[j]) {
					j++
				}
				if j >= len(raw) {
					i = len(raw)
					break
				}
				body := raw[start:j]
				final := raw[j]
				private := body != "" && strings.ContainsRune("?<>=", rune(body[0]))
				nums := parseParams(strings.TrimLeft(body, "?<>="))

				switch {
				case final == 'm' && !private:
					flush()
					style.apply(nums)
					out.Controls.SGR++
				case final == 'K':
					out.Controls.EraseLine++
				case final == 'J':
					out.Controls.EraseDisplay++
				case strings.IndexByte("ABCDEFGHfd", final) >= 0:
					out.Controls.CursorMove++
				case private && (final == 'h' || final == 'l'):
					mode := 0
					if len(nums) > 0 {
						mode = nums[0]
					}
					if mode == 25 {
						if final == 'l' {
							out.Controls.CursorHide++
						} else {
							out.Controls.CursorShow++
						}
					}
					if mode == 1049 || mode == 47 || mode == 1047 {
						if final == 'h' {
							out.Controls.AltScreenEnter++
						} else {
							out.Controls.AltScreenLeave++
						}
					}
				}
				i = j
				continue
			}
			if i+1 < len(raw) && raw[i+1] == ']' {
				j := i + 2
				for j < len(raw) && raw[j] != 0x07 && !(raw[j] == 0x1b && j+1 < len(raw) && raw[j+1] == '\\') {
					j++
				}
				out.Controls.OSC++
				if j < len(raw) && raw[j] == 0x1b {
					j++
				}
				i = j
				continue
			}
			i++
			continue
		}

		if ch == '\r' {
			out.Controls.CarriageReturns++
			plain.WriteByte(ch)
			continue
		}
		if ch == 0x07 {
			out.Controls.Bell++
			continue
		}

		plain.WriteByte(ch)
		if ch == '\n' {
			flush()
			lineNum++
			continue
		}

		if cur == nil {
			s := style
			s.FG = EffectiveFG(style)
			cur = &Run{Style: s, Line: lineNum}
		}
		cur.Text += string(ch)
	}
	flush()

	out.Plain = plain.String()
	return out
}

func parseParams(body string) []int {
	if body == "" {
		return nil
	}
	parts := strings.Split(body, ";")
	nums := make([]int, 0, len(parts))
	for _, p := range parts {
		if p == "" {
			nums = append(nums, 0)
			continue
		}
		if n, err := strconv.Atoi(p); err == nil {
			nums = append(nums, n)
		}
	}
	return nums
}

// Strip removes every escape sequence, leaving what a plain-text reader gets.
func Strip(raw string) string { return Parse(raw).Plain }

var sgrRe = regexp.MustCompile(`\x1b\[[0-9;]*m`)

// IsColourParam reports whether an SGR parameter actually changes a colour.
// Bold, underline and italic are emphasis: they survive a monochrome terminal,
// they carry meaning without it, and NO_COLOR is not asking anyone to give them
// up. Faint and reverse do change what colour lands on the screen, so they
// count.
func IsColourParam(n int) bool {
	return (n >= 30 && n <= 49) || (n >= 90 && n <= 97) || (n >= 100 && n <= 107) ||
		n == 2 || n == 7
}

// HasColour reports whether a stream sets a colour, as opposed to merely
// emphasising text.
//
// The distinction matters more than it looks. A pager renders man-page style
// help with bold and underline, and those bytes arrive in a capture mixed in
// with the CLI's own; counting them as colour means accusing a perfectly
// well-behaved tool of ignoring NO_COLOR.
func HasColour(raw string) bool {
	for _, seq := range sgrRe.FindAllString(raw, -1) {
		for _, n := range parseParams(seq[2 : len(seq)-1]) {
			if IsColourParam(n) {
				return true
			}
		}
	}
	return false
}

// CountSGR counts styling sequences, for reporting.
func CountSGR(raw string) int { return len(sgrRe.FindAllString(raw, -1)) }
